chia/util/files.py

```diff
@@ -0,0 +1,88 @@
+import asyncio
+import logging
+import os
+import shutil
+
+from aiofiles import tempfile  # type: ignore
+from pathlib import Path
+from typing import Union
+
+
+log = logging.getLogger(__name__)
+
+
+def move_file(src: Path, dst: Path):
+    """
+    Attempts to move the file at src to dst, falling back to a copy if the move fails.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(dst.parent, mode=dir_perms, exist_ok=True)
+
+    try:
+        # Attempt an atomic move first
+        os.replace(os.fspath(src), os.fspath(dst))
+    except Exception as e:
+        log.debug(f"Failed to move {src} to {dst} using os.replace, reattempting with shutil.move: {e}")
+        try:
+            # If that fails, use the more robust shutil.move(), though it may internally initiate a copy
+            shutil.move(os.fspath(src), os.fspath(dst))
+        except Exception:
+            log.exception(f"Failed to move {src} to {dst} using shutil.move")
+            raise
+
+
+async def move_file_async(src: Path, dst: Path, *, reattempts: int = 6, reattempt_delay: float = 0.5):
+    """
+    Attempts to move the file at src to dst, making multiple attempts if the move fails.
+    """
+
+    remaining_attempts: int = reattempts
+    while True:
+        try:
+            move_file(src, dst)
+        except Exception:
+            if remaining_attempts > 0:
+                log.debug(f"Failed to move {src} to {dst}, retrying in {reattempt_delay} seconds")
+                remaining_attempts -= 1
+                await asyncio.sleep(reattempt_delay)
+            else:
+                break
+        else:
+            break
+
+    if not dst.exists():
+        raise FileNotFoundError(f"Failed to move {src} to {dst}")
+    else:
+        log.debug(f"Moved {src} to {dst}")
+
+
+async def write_file_async(file_path: Path, data: Union[str, bytes], *, file_mode: int = 0o600):
+    """
+    Writes the provided data to a temporary file and then moves it to the final destination.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(file_path.parent, mode=dir_perms, exist_ok=True)
+
+    mode: str = "w+" if type(data) == str else "w+b"
+    temp_file_path: Path
+    async with tempfile.NamedTemporaryFile(dir=file_path.parent, mode=mode, delete=False) as f:
+        temp_file_path = f.name
+        await f.write(data)
```

Comment:
I think it would be a good idea to add:
```
await f.flush()
```

here, to make sure the data actually hits the disk before swapping the files out, in case of catastrophic failures at the wrong time

Reply:
I don't know that it would make a difference, but maybe. Since this is used in a context manager via `async with` the temp file will be closed before the files are swapped out and any buffers should be flushed in the process. Unless the close is a lie.

Reply:
the [documentation](https://docs.python.org/3/library/io.html#io.IOBase.close) says:

> Flush and close this stream

And the documentation for `flush()` says:

> Flush the write buffers of the stream if applicable.

So, it's not explicit about whether it's also blocks until that flush completes, but it seems reasonable to assume that.

Reply:
I'll add a call to flush to make it explicit.